[Polkadot Wiki Migration] Offenses and Slashes (#43)
polkadot-developers/polkadot-docs: infrastructure/staking-mechanics/offenses-and-slashes.md

- **Equivocation** - a validator produces two or more identical blocks or votes
- **GRANDPA and BEEFY equivocation** - a validator signs two or more votes in the same round on different chains
- **BABE equivocation** - a validator produces two or more blocks on the relay chain in the same time slot
- **Double seconded equivocation** - in a backing group of five para-validators, five backed parablocks are possible. Each parablock requires one seconded vote and at least two valid votes from the five potential backers, establishing an upper limit on the number of parablocks the system can manage while providing some flexibility for relay chain block authors. Backers must choose one parablock to second; seconding more than one incurs punishment. [Asynchronous backing](https://wiki.polkadot.network/docs/learn-async-backing){target=\_blank} complicates this by allowing backers to support blocks "into the future" optimistically rather than being restricted to one candidate per relay chain block. See the documentation on [Seconding Limit](https://paritytech.github.io/polkadot-sdk/book/node/backing/statement-distribution.html#seconding-limit){target=\_blank} for more information

> - **Seconded and valid equivocation** - a malicious node seconds a vote, taking absolute responsibility for it, and then falsely claims it is correct after someone else has taken responsibility. Once conflicting votes are detected by the system, the offense is reported

Reviewer: I'm not in love with this because I think `ForInvalid` and `AgainstValid` are very specific things in the code, but I'm not sure about the other items in this list. I don't know enough here and tried looking into it a little bit but opted out of going down the rabbit hole haha, so I'll defer to @polkadot-developers/docs-content-team on this 🙂

Reviewer: Also, I added some formatting to this list, but not the table. This will ideally in the future be formatted the same across the board, but I want to better understand the specifics of the other items first

Author: So from my knowledge (which is not a lot in this specific area, @filippoweb3, want to give this a shot?), these do all have references in code, whether it is an enum or a pallet specific function which specifies this offense (as is the case with BABE)

You are correct in that they references specific things, see here: https://paritytech.github.io/polkadot-sdk/master/cumulus_primitives_core/relay_chain/v8/slashing/enum.SlashingOffenceKind.html#

Perhaps we could rephrase them, like: "`ForInvalid`..." -> "Invalid vote"... to match the rest, or we could rename all of them, not sure which is better

Reviewer 2: @Overkillus might have some insights here. We worked on the Wiki page about offenses this year. Unsure if a duplicate is needed on the devs docs. You could just refer to the Wiki page.

Reviewer 3: @filippoweb3  - Ideally we move everything that is dev oriented (including Infra) from the Wiki to the Devs docs.

I would love to connect with him via Telegram as well so that we can speed this up 😃

Reviewer 4: I did some work on this page to change a bit how it's laid out and how the information is presented. I'd love to have both teams take a look and let me know what you think!

---
title: Offenses and Slashes
description: Learn about how Polkadot discourages validator misconduct via an offenses and slashing system, including details on offenses and their consequences.
---

# Offenses and Slashes

## Introduction

> In Polkadot's Nominated Proof-of-Stake (NPoS) system, validator misconduct is deterred through a combination of slashing, deactivation, and reputation penalties. Validators and nominators who stake tokens face consequences for validator misbehavior, which range from token slashes to restrictions on network participation.

Reviewer: ```suggestion
In Polkadot's Nominated Proof of Stake (NPoS) system, validator misconduct is deterred through a combination of slashing, deactivation, and reputation penalties. Validators and nominators who stake tokens face consequences for validator misbehavior, which range from token slashes to restrictions on network participation.
```

This page outlines the types of offenses recognized by Polkadot, including block equivocations and invalid votes, as well as the corresponding penalties. While some parachains may implement additional custom slashing mechanisms, this guide focuses on the offenses tied to staking within the Polkadot ecosystem.

> ??? info "Deactivation"

Reviewer: I'd say this seems important and we should expand it

Reviewer: This also feels a bit weird to be here as a note without any mentioning of it beforehand

    The deactivation mechanism is triggered when validators commit serious infractions, such as backing invalid blocks or engaging in equivocations. The material in this guide reflects the changes introduced in Stage 2. For more details, refer to the [State of Disabling issue](https://github.com/paritytech/polkadot-sdk/issues/4359){target=\_blank} on GitHub.

## Offenses

> Polkadot is a public permissionless network. As such, it has a mechanism to disincentivize offenses and incentivize good behavior. You can review the [parachain protocol](https://wiki.polkadot.network/docs/learn-parachains-protocol#parachain-protocol){target=\_blank} to better understand the terminology used to describe offenses. Polkadot validator offenses fall into two categories, invalid votes and equivocations. 

Reviewer: ```suggestion
Polkadot is a public permissionless network. As such, it has a mechanism to disincentivize offenses and incentivize good behavior. You can review the [parachain protocol](https://wiki.polkadot.network/docs/learn-parachains-protocol#parachain-protocol){target=\_blank} to understand better the terminology used to describe offenses. Polkadot validator offenses fall into two categories: invalid votes and equivocations. 
```

    - **BABE equivocation** - the validator produces two or more blocks on the relay chain in the same time slot
- **Double seconded equivocation** - the validator attempts to second, or back, more than one block in the same round
- **Seconded and valid equivocation** - the validator seconds, or backs, a block and then attempts to hide their role as the responsible backer by later placing a standard validation vote

## Penalties

On Polkadot, offenses to the network incur different penalties depending on severity. There are three main penalties: slashing, deactivation, and reputation changes.

### Slashing

Validators engaging in bad actor behavior in the network may be subject to slashing if they commit a qualifying offense. When a validator is slashed, they and their nominators lose a percentage of their staked DOT or KSM, from as little as 0.01% up to 100% based on the severity of the offense. Nominators are evaluated for slashing against their active validations at any given time. Validator nodes are evaluated as discrete entities, meaning an operator can't attempt to mitigate the offense on another node they operate in order to avoid a slash. 

> Any slashed DOT or KSM will be added to the [Treasury](https://wiki.polkadot.network/docs/learn-polkadot-opengov-treasury){target=\_blank} rather than burned or distributed as rewards. Moving slashed funds to the Treasury allows an opportunity to quickly move tokens away from malicious validators while maintaining the ability to revert faulty slashes when needed.

Reviewer: ```suggestion
Any slashed DOT or KSM will be added to the [Treasury](https://wiki.polkadot.network/docs/learn-polkadot-opengov-treasury){target=\_blank} rather than burned or distributed as rewards. Moving slashed funds to the Treasury allows tokens to be quickly moved away from malicious validators while maintaining the ability to revert faulty slashes when needed.
```

??? info "Multiple active nominations"
    A nominator with a very large bond may nominate several validators in a single era. In this case, a slash is proportionate to the amount staked to the offending validator. Stake allocation and validator activation is controlled by the [Phragmén algorithm](https://wiki.polkadot.network/docs/learn-phragmen#understanding-phragm%C3%A9n){target=\_blank}.

> A validator slash creates an `unapplied` state transition. You can view pending slashes on [Polkadot.js Apps](https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Frpc.polkadot.io#/staking/slashes){target=\_blank}. The UI will display the slash per validator, followed by the affected nominators and the slash amounts. The unapplied state includes a 27-day grace period during which a governance proposal can be made to reverse the slash. Once this grace period expires, the slash is applied.

Reviewer: ```suggestion
A validator slash creates an `unapplied` state transition. You can view pending slashes on [Polkadot.js Apps](https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Frpc.polkadot.io#/staking/slashes){target=\_blank}. The UI will display the slash per validator, the affected nominators, and the slash amounts. The unapplied state includes a 27-day grace period during which a governance proposal can be made to reverse the slash. Once this grace period expires, the slash is applied.
```

#### Equivocation Slash

> The Web3 Foundation's [Slashing mechanisms](https://research.web3.foundation/Polkadot/security/slashing/amounts){target=\_blank} page provides guidelines for evaluating the security threat level of different offenses and determining penalties proportionate to the threat level of the offense. Offenses requiring coordination between validators or extensive computational costs to the system will typically call for harsher penalties than those more likely to be unintentional than malicious. A description of potential offenses for each threat level, and the corresponding penalties, is as follows:
> 
> - **Level 1** - honest misconduct such as isolated cases of unresponsivness
>     - **Penalty** - validator can be kicked out or slashed up to 0.1% of stake in the validator slot
> - **Level 2** - misconduct that can occur honestly but is a sign of bad practices. Examples include repeated cases of unresponsiveness and isolated cases of equivocation
>     - **Penalty** - slash of up to 1% of stake in the validator slot
> - **Level 3** - misconduct that is likely intentional but of limited affect on the performance or security of the network. This level will typically include signs of coordination between validators. Examples include repeated cases of equivocation or isolated cases of unjustified voting on GRANDPA
>     - **Penalty** - reduction in networking reputation metrics, slash of up to 10% of stake in the validator slot
> - **Level 4** - misconduct that poses severe security or monetary risk to the system or mass collusion. Examples include signs of extensive coordination, creating a serious security risk to the system, or forcing the system to use extensive resources to counter the misconduct
>     - **Penalty** - slash of up to 100% of stake in the validator slot

Reviewer: ```suggestion
The Web3 Foundation's [Slashing mechanisms](https://research.web3.foundation/Polkadot/security/slashing/amounts){target=\_blank} page provides guidelines for evaluating the security threat level of different offenses and determining penalties proportionate to the threat level of the offense. Offenses requiring coordination between validators or extensive computational costs to the system will typically call for harsher penalties than those more likely to be unintentional than malicious. A description of potential offenses for each threat level and the corresponding penalties is as follows:

- **Level 1** - honest misconduct such as isolated cases of unresponsiveness
    - **Penalty** - validator can be kicked out or slashed up to 0.1% of stake in the validator slot
- **Level 2** - misconduct that can occur honestly but is a sign of bad practices. Examples include repeated cases of unresponsiveness and isolated cases of equivocation
    - **Penalty** - slash of up to 1% of stake in the validator slot
- **Level 3** - misconduct that is likely intentional but of limited effect on the performance or security of the network. This level will typically include signs of coordination between validators. Examples include repeated cases of equivocation or isolated cases of unjustified voting on GRANDPA
    - **Penalty** - reduction in networking reputation metrics, slash of up to 10% of stake in the validator slot
- **Level 4** - misconduct that poses severe security or monetary risk to the system or mass collusion. Examples include signs of extensive coordination, creating a serious security risk to the system, or forcing the system to use extensive resources to counter the misconduct
    - **Penalty** - slash of up to 100% of stake in the validator slot
```

See the next section to understand how slash amounts for equivocations are calculated. If you want to know more details about slashing, please look at the research page on [Slashing mechanisms](https://research.web3.foundation/Polkadot/security/slashing/amounts){target=\_blank}.

#### Slash Calculation for Equivocation

The slashing penalty for GRANDPA, BABE, and BEEFY equivocations is calculated using the formula below, where `x` represents the number of offenders and `n` is the total number of validators in the active set:

```text
min((3 * x / n )^2, 1)
```

The following scenarios demonstrate how this formula means slash percentages can increase exponentially based on the number of offenders involved compared to the size of the validator pool:

> - **Minor offense** - assume one validator out of a 100 validator active set equivocates in a slot. A single validator committing an isolated offense is most likely a mistake rather than malicious attack on the network. This offense results in a 0.09% slash to the stake in the validator slot

Reviewer: ```suggestion
- **Minor offense** - assume 1 validator out of a 100 validator active set equivocates in a slot. A single validator committing an isolated offense is most likely a mistake rather than malicious attack on the network. This offense results in a 0.09% slash to the stake in the validator slot
```

    ``` mermaid
    flowchart LR
    N["Total Validators = 100"]
    X["Offenders = 1"]
    F["min(3 * 1 / 100)^2, 1) = 0.0009"]
    G["0.09% slash of stake"]

    N --> F
    X --> F
    F --> G
    ```

> - **Moderate offense** - assume five validators out a 100 validator active set equivocate in a slot. This is a slightly more serious event as there may be some element of coordination involved. This offense results in a 2.25% slash to the stake in the validator slot

Reviewer: ```suggestion
- **Moderate offense** - assume 5 validators out a 100 validator active set equivocate in a slot. This is a slightly more serious event as there may be some element of coordination involved. This offense results in a 2.25% slash to the stake in the validator slot
```

#### Slashing Across Eras

There are three main difficulties to account for with slashing in NPoS:

- A nominator can nominate multiple validators and be slashed as a result of actions taken by any of them
- Until slashed, the stake is reused from era to era
- Slashable offenses can be found after the fact and out of order

To balance this, the system applies only the maximum slash a participant can receive in a given time period rather than the sum. This ensures protection from excessive slashing.

### Deactivation

Deactivation stops validators from performing specific actions after they have committed an offense. Deactivation is further divided into:

- **On-chain deactivation** - lasts for a whole era and stops validators from authoring blocks, backing, and initiating a dispute
- **Off-chain deactivation** - lasts for a session, is caused by losing a dispute, and stops validators from initiating a dispute

Off-chain deactivation is always a lower priority than on-chain deactivation. Off-chain deactivation prioritizes deactivating first backers and then approval checkers.

### Reputation Changes

> Some minor offenses, such as spamming, are only punished by networking reputation changes. Validators use a reputation metric when chosing with which peers to connect. If a peer provides valuable data and behaves appropriately, the system adds reputation. If they provide faulty or spam data, the system reduces their reputation. If a validator loses enough reputation their peers will temporarily close their channels to them. This helps in fighting against DoS (Denial of Service) attacks. In general, performing validator tasks under reduced reputation will be harder, resulting in lower validator rewards.

Reviewer: ```suggestion
Some minor offenses, such as spamming, are only punished by networking reputation changes. Validators use a reputation metric when choosing which peers to connect with. The system adds reputation if a peer provides valuable data and behaves appropriately. If they provide faulty or spam data, the system reduces their reputation. If a validator loses enough reputation, their peers will temporarily close their channels to them. This helps in fighting against Denial of Service (DoS) attacks. Performing validator tasks under reduced reputation will be harder, resulting in lower validator rewards.
```

---
title: Offenses and Slashes
description: Learn about how Polkadot discourages validator misconduct via an offenses and slashing system, including details on offenses and their consequences.
---

# Offenses and Slashes

## Introduction

> In Polkadot's Nominated Proof-of-Stake (NPoS) system, validator misconduct is deterred through a combination of slashing, deactivation, and reputation penalties. Validators and nominators who stake tokens face consequences for validator misbehavior, which range from token slashes to restrictions on network participation.

Reviewer 4: I was trying to bring the language in line with inclusivity standards (here's Google style guide take: https://developers.google.com/style/inclusive-documentation). I will change it back if it really needs to be but I won't like it ;-)

> In Polkadot's Nominated Proof-of-Stake (NPoS) system, validator misconduct is deterred through a combination of slashing, deactivation, and reputation penalties. Validators and nominators who stake tokens face consequences for validator misbehavior, which range from token slashes to restrictions on network participation.

Reviewer: I hear ya, but if that's the name of the feature, I sort of feel like our hands are tied a bit
